### ISSUE TITLE
Base Failure Rate appears far too high to be viable for space missions.

### ISSUE THREAD

**Issue by diomedea** [KSP-RO/TestFlight#81]
Since the first time a part is activated, Flight Data are collected up to the maxData value. Currently, maxData is set to 10000 du for every part; even with the smallest flightDataMultiplier currently used (=2) that means 10000/2 = 5000 frames to collect maxData data units (even with a rather slow computer able to run 10 frames/second, that means 500 seconds). Sure maxData is configurable, but the span of time needed to complete some space missions hints at a problem with any specific duration
(e.g., to perform a simple Kerbin-Jool Hohmann transfer, lowest possible DV, the time of flight is 385.4 Earth days ~= 33 M seconds).
Now, the collected du allow the TestFlightReliability module to define the Base Failure Rate (BFR). At a du value = maxData the BFR is the lowest. The lowest value charted in a config for BFR is (for KW1mengineMaverick1D) = 0.000002. That failure rate is sure very low, but when cumulated over 33 M seconds, the cumulative failure probability (CFP) is 1-((1 - 0.000002)^33M) = 99.999...% (add 27 more 9s after the decimal point). And here is the problem: space flight duration requires a lot lower failure rate to be acceptable. In reality, if any essential part had a CFP of more than 1%, the mission would be considered risky. With the rather slacky risk management used by kerbals, I may imagine they running missions with 10% CFP (For a CFP of 10% and a Flight Time of 33M seconds, the required BFR has to be 3.2E-9, 4 orders of magnitude lower than the actual minimum).
Needless to say, with a BFR as low as the one shown above, no failure would occur in those critical short moments of a mission when a failure is more probable.

Now, the above brings to some further considerations, with the intent to find a solution. 
Failures in TestFlight can occur only after a part is activated. Unfortunately, if a vessel is designed with a stage for mid-course corrections while enroute to Jool, such a stage has to be made active and will almost certainly have failed before Jool is reached. If was possible to "de-activate" parts so TestFlight failures would not occur, that may be used as a viable strategy (however, shutting down engines doesn't stop TestFlight, and tanks have no mechanic to stop them being considered for failures). And, other parts not yet defined in TestFlight configs (command modules, batteries, solar panels, antennae) have to be active at all times so won't be spared.

The occurrence of failures may instead be tied to some specific conditions. The EngineCycle module already provides failures specifically about running engines. Other modules may be conceived to handle failures when stressful conditions exist, like: overheating, high pressure, splashed, or within a "corrosive" atmosphere. Given that such specific conditions should not be the norm in space, the failures those define would not happen in orbital flights, and a choice of an extremely low BFR (such as the 3.2E-9 shown above, or even lower) would make sense.

**Comment by jwvanderbeck** [comment]
A couple of remarks really quick, then i'll read over this in more detail
when I have some free time.

1. Data Rate is scaled with time, not per frame.  This was it is consistent
across machines, and works as expected in time warp.
2. The existing rates on the engines were based on the rated burn times.
Basically assuming an engine lasts until roughly its rated burn time, you
will acquire maximum data in about 10-15 flights if I recall correctly
3. Once an engine is shutdown, it should no longer be gathering data nor
should it have the ability to fail.  If an engine that is shut down is
still failing after time, please open a separate issue for that

The bulk of time spent in TestFlight has been on handling engines, so
handling the other parts is something that I just need to take deal with as
they come up.  I fully agree with special "cycle" style modules for other
part types if they make sense.

- John Vanderbeck
- http://www.johnvanderbeck.com

**Comment by jwvanderbeck** [comment]
Just to follow up on BFR and what not, what you are seeing on the engines is kind of the intention behind how other parts should be handled  A very very low BFR that gives MTBF times of like months or years, but then special more part specific modules that cause the MFR, or momentary failure rate to increase.  In the case of engines it is a cycle thing where it simulates the points of an engine cycle more likely to fail.  Other parts can, and should, be treated similarly it is just a matter of 1) Defining how that should work, and 2) Writing the modules to handle it.

I started with engines because they were both more complex, and at the same time more simple in that I knew what sorts of stresses would be likely simulation targets.  Things like tanks though, I am not so sure on.  Maybe a cycle module that has an occasional blip in MFR to simulate something like cryo stirring?  I'm not sure.

**Comment by diomedea** [comment]
Lots of good points, thanks for those comments above.
1. Appreciate that Data Rate scales with time, it is certainly the correct way. Believe I got confused about the checks occurring at each frame.
2. Engines are really the parts better defined, having the reliability curves  and the Failure Rates closest to reality. Indeed, both Data collection and Failure checkings have only to occur when engines are used. 
3. Engines not correctly configured keep doing the above (Data collection, Failure checking) when shut. However, I found that related to engines defined in add-ons (not stock, not RO), if they possess any resource (e.g. EC) so are receiving the same behaviour of tanks. Should not be worth an issue, but will have to make configs for those.

Indeed, at the moment tanks (any part containing a resource) have that disproportionate risk of failure given enough time. Engines properly configured show what can be achieved instead. When burning, engines are under stress, the EngineCycle module really captures this condition effectively. But other parts also need some specific module to manage the MFR, in relation to what stresses those parts experience. BFR alone can't go much further than simulate events unrelated to stress (yes, ruptures due to cryotemperatures, the occasional gamma ray burst burning electronics, or a micrometeorite piercing a tank).

**Comment by jwvanderbeck** [comment]
Yeah we're thinking along the same lines.  Really it is just a matter of we need more modules to do things :)  And that is fine.

Just need the ideas on what to do :)

**Comment by stratochief66** [comment]
There are a number of factors to be balances in this equation, chiefly realism vs mini-game/playability.

It is certainly realistic to have part failures relatively often for low and mid-tech engines, but does it add to play? Some (more hardcore than me) would say yes, but I would propose an easier play mode for the less hardcore.

Perhaps when lower difficulty level is set, engines could experience loss of thrust and/or loss of efficiency events in place of complete engine failures. This would make failures something to try to anticipate and plan for, rather than completely scrubbing the mission. This would also encourage BARIS type behaviour, where a mission can effectively be downgraded due to a failure. If the engine is no longer efficient enough to perform a lunar flyby, RealScience experiments could be performed in medium earth orbit instead!

Just a thought.

**Comment by diomedea** [comment]
I had a direct confirmation that Tanks BFR needs a review (using the stock Tanks config, of course). Tried one mission to MInmus, within the 9 days required to arrive in MInmus' SoI, every single one of the tanks (batteries included) had leaked their contents. Having only the BFR to tweak currently, I'm now reducing that from the 2*10E-5 value (for the max amount of du with all stock tanks) to something like the 2*10E-9 (coherent with the probabilitity of failure over a long mission). Unfortunately that lowers the failure probability much below the MFR in stressful situations, but can't let even the easiest missions to fail within few days.

**Comment by icedown** [comment]
How about adding a max/min data or fail rate that failures can happen at.  While I do realize this isn't exactly realistic, an almost guaranteed total failure of any interplanetary probe is not as well.   You could use a max for saying that after this much data(or this low of failure rate) that this tank will no longer explode.  With a min in conjunction with a heavily weighted bug type failure(it may say it's leaking, but it's really not, and can be repaired easily), you could set that after a certain amount of data has been gathered, it allows this type of failure to happen.  This would greatly reduce the amount of spacecraft crippling failures.  Another solution, that might be on it's own or go along with this is a scaling of the capability to repair the part.

**Comment by jwvanderbeck** [comment]
If you can, please test out the latest development build: http://www.johnvanderbeck.com/projects/testflight/dev/25

That attempts to deal with this issue.
